### PR TITLE
Add aggregation modules for different scenarios.

### PR DESCRIPTION
**Pull request by LiamHe**
For the moment, we use something like the following snippet for communication and applying gradient.
```python
# loss.backward()
# Average the gradients from all of the nodes
global_average(param.grad.data, op=dist.op.SUM)
# Apply gradients to weights
optimizer.step()
```
But this is not flexible enough when we need to
- average with neighbors instead of all nodes;
- average the weight instead of gradient;
- communicate a sparsified / quantized vector instead of full vector;
- ...

In this PR, the control flow is changed to remove the explicit `global_average` and simply use the following snippet to update gradient
```python
# loss.backward()
# Apply gradients to weights
optimizer.step()
```
It is up to user to decide how to apply the gradients from all of the nodes. If the optimizer is native `optim.SGD`, then the `.step()` does not average the gradients. We can simply use a  wrapper or a subclass of `optim.SGD` and overwrite `.step()` to take care of communication.

Associated examples (decentralized, sign sgd, sparsified sgd etc) will be added to mlbench-benchmarks. 

Besides, a few changes have been made like
- Fix bugs in resuming training from checkpoint;
- Changes for flake8 linting information.

### CODE REVIEW

File: mlbench_core/controlflow/pytorch/controlflow.py [hunk 1]
```diff
@@ -123,40 +123,39 @@ def run(self, dataloader_train=None, dataloader_val=None,
...
-                            self.timeit.cumu))
+            logger.info("Current epoch : {} : lr={}"
+                        .format(epoch, self.scheduler.get_lr()))
 
             # FIXME: The Timeit object can be a problem.
             self.train_epoch(dataloader_train)
 
-            if self.perform_validation:
-                self.do_validate(dataloader_val)
+            is_best = self.do_validate(
```
Comment (Panaetius) [inline mlbench_core/controlflow/pytorch/controlflow.py]:
```
is_best = False
if self.perform_validation:
    is_best = self.do_validate(dataloader_val)
```
Is more readable i.e. it's harder to miss what the line is actually doing.

File: mlbench_core/lr_scheduler/pytorch/lr.py [hunk 2]
```diff
@@ -155,6 +163,54 @@ def f(duration):
...
+    def __init__(self,
+                 optimizer,
+                 world_size,
+                 gamma,
+                 milestones,
+                 lr,
+                 warmup_duration,
+                 warmup_linear_scaling=True,
+                 warmup_init_lr=None):
+        if not list(milestones) == sorted(milestones):
```
Comment (Panaetius) [inline mlbench_core/lr_scheduler/pytorch/lr.py]: Use `!=` instead

File: mlbench_core/models/pytorch/resnet.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def batch_norm(num_features):
     """Create a batch normalization layer.
 
     See the Disclaimers in Kaiming's
-    `github repository <https://github.com/KaimingHe/deep-residual-networks/tree/a7026cb6d478e131b765b898c312e25f9f6dc031>`_
+    `here <https://github.com/KaimingHe/deep-residual-networks/tree/a7026cb6d478e131b765b898c312e25f9f6dc031>`_
```
Comment (Panaetius) [inline mlbench_core/models/pytorch/resnet.py]: "See the disclaimer in Kaiming's here" <-- something more descriptive than "here" would be good :) or did you forget to also delete "In Kaiming's" ?

File: mlbench_core/utils/pytorch/checkpoint.py [hunk 4]
```diff
@@ -1,37 +1,46 @@
...
         self.rank = rank
-        self.checkpoint_all = checkpoint_all
-        self.runtime = {'cumu_time_val': []}
+        self.freq = freq
+        self.save_stats = save_stats
+        # self.runtime = {'cumu_time_val': []}
 
-    def get_ckpt_id(self, epoch):
-        """ Get the name of a checkpoint
+    # def get_ckpt_id(self, epoch):
```
Comment (Panaetius) [inline mlbench_core/utils/pytorch/checkpoint.py]: Can this commented code be removed?

File: mlbench_core/utils/pytorch/checkpoint.py [hunk 5]
```diff
@@ -51,19 +60,35 @@ def save(self, tracker, model, optimizer, scheduler, epoch, is_best):
...
-        if self.checkpoint_all:
-            torch.save(state, checkpoint_path)
+        if self.freq == CheckpointFreq.ALL:
+            torch.save(state, checkpoint_path, pickle_module=dill)
             if is_best:
                 shutil.copyfile(checkpoint_path, best_model_path)
-        else:
+        elif self.freq == CheckpointFreq.BEST:
             torch.save(state, best_model_path, pickle_module=dill)
+        elif self.freq == CheckpointFreq.NONE:
```
Comment (Panaetius) [inline mlbench_core/utils/pytorch/checkpoint.py]:
```
elif self.freq == CheckpointFreq.NONE:
            pass
        else:
            raise NotImplementedError
```

can be simplified to
```
elif self.freq != CheckpointFreq.NONE:
            raise NotImplementedError
```

File: mlbench_core/utils/pytorch/checkpoint.py [hunk 6]
```diff
@@ -111,23 +181,24 @@ def determine_restore_ckpt_path(rank, checkpoint_root):
...
 
-def maybe_resume(config, model, optimizer, scheduler):
-    """Recover the state of config, model, optimizer and scheduler."""
-    if 'resume' in config and config['resume']:
-        # reload model from the latest checkpoint.
-        config['runtime'] = resume(config, model, optimizer, scheduler)
-    else:
-        config['runtime'] = {}
-    return config
+# def maybe_resume(config, model, optimizer, scheduler):
```
Comment (Panaetius) [inline mlbench_core/utils/pytorch/checkpoint.py]: Can this be removed?

File: mlbench_core/utils/pytorch/helpers.py [hunk 7]
```diff
@@ -177,7 +172,7 @@ def config_pytorch(use_cuda=False, seed=None, cudnn_deterministic=False):
     return rank, world_size, graph
 
 
-def log_metrics(run_id, rank, epoch, metric_name, value):
+def log_metrics(run_id, rank, epoch, metric_name, value, tracker=None, time=None):
```
Comment (Panaetius) [inline mlbench_core/utils/pytorch/helpers.py]: log_metrics is deprecated, change https://github.com/mlbench/mlbench-core/blob/develop/mlbench_core/utils/pytorch/helpers.py#L182 instead.
log_metrics has performance implications since it can take quite some time (up to 5% of all running time) to create the ApiClient, `LogMetrics` uses a cached ApiClient (this only uses 0.1% of total runtime in my experiments).
